### PR TITLE
Improve checking the file's status before shredding

### PR DESCRIPTION
**Pull request by githubuserx**
Improve checking a file's status by

- clearing PHP's [stat cache](https://secure.php.net/manual/de/function.clearstatcache.php) for the specific file
- checking if it actually is a file, otherwise could be a directory

### CODE REVIEW

File: src/Shred.php [hunk 1]
```diff
@@ -98,7 +98,13 @@ public function shred($filepath, $remove = true) {
    * @return bool
    */
   private function fileWritable($filepath) {
-    if (is_readable($filepath) && is_writable($filepath)) {
+
+    // clear stat cache to avoid falsely reported file status
+    // use $filename parameter to possibly improve performance
+    clearstatcache(true, $filepath);
```
Comment (DanielRuf) [inline src/Shred.php]: Should be probably also done after the `unlink()` call or before the `return $unlink`.

But

> This function caches information about specific filenames, so you only need to call clearstatcache() if you are performing multiple operations on the same filename and require the information about that particular file to not be cached.

Also I think the other is already covered Let me add a unit test which tests for directories and non existing files.

Comment (githubuserx) [reply src/Shred.php]: >  However unlink() clears the cache automatically.

Please also see
https://secure.php.net/manual/de/function.clearstatcache.php#105888

It maybe could be useful though to clear the cache again *before* deleting the file (`unlink()`), as in theory another process could have accessed it after the initial check.

I don't think checking if it's a file is covered anywhere else, but `file_exists()` should be covered by `is_readable()`.

Comment (DanielRuf) [reply src/Shred.php]: > Should be probably also done after the `unlink()` call or before the `return $unlink`.

There are file modifications done which are probably relevant (mdate for example).
Let's better check this with proper unit tests.

Comment (DanielRuf) [reply src/Shred.php]: Can you add a few unit tests to cover these cases?
If not I'll try to find some time to provide some.

Comment (githubuserx) [reply src/Shred.php]: Can you elaborate on what exactly you want to test for?
I'm also not sure why mdate would be relevant.

For now I wrote a unit test regarding directories, which fails on the `master` branch with
```
Exception : 0 :: Cannot use SplFileObject with directories ::
```
whereas with the changes of this PR, `shred()` would return `false` as I'd expect it.

Furthermore, I'm not sure if the behaviour of the stat cache can be properly tested with `vfsStreams`.

**Comment by DanielRuf** [comment]
Can you rebase? Also if there are any unit tests for this PR, please add them so I can merge them with the PR.

File: src/Shred.php [hunk 2]
```diff
@@ -98,7 +98,14 @@ public function shred($filepath, $remove = true) {
...
+
+    if (is_file($filepath)
+      && is_readable($filepath)
+      && is_writable($filepath)) {
```
Comment (DanielRuf) [inline src/Shred.php]: Not sure if these are sufficient for users to know that there is an issue (if it is not shredded) or if we shall throw Exceptions (disabled by default). What do you think?

Would be interesting for another PR.

Comment (githubuserx) [reply src/Shred.php]: As the usage instructions suggest that users should check the return value and passing a directory is wrong usage, I think returning false is enough in this case.
I would only throw exceptions if something goes wrong after the actual shredding started, since then it is possible that the file is left in an inconsistent state and manual user intervention could be necessary to delete e.g. a remaining file.

**Comment by DanielRuf** [comment]
Thank you very much for your contribution and improving this library.
